Rename 'io.jaegertracing.Span' to 'io.jaegertracing.JaegerSpan'

jpkrohling opened the pull request:
## Which problem is this PR solving?
- Resolves #434

## Short description of the changes
* Rename Jaeger's `Span` to `JaegerSpan`
* Rename Jaeger's `SpanContext` to `JaegerSpanContext`
* Rename Jaeger's `Tracer` to `JaegerTracer`

Not included in this PR are the decisions about what should be public and what should be internal API. This will be handled in #396.

yurishkuro commented:
can we avoid the name JaegerBaseSpan?  I don't understand why we cannot use packages for namespacing.

pavolloffay commented on jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java:
```diff
@@ -0,0 +1,43 @@
+ * Should be used wisely by instrumented applications: always prefer OpenTracing's approach when available.
+ */
+public interface JaegerSpan extends Span {
+  Tracer getTracer();
```
I would prefer not expose tracer here.

jpkrohling replied:
I extracted `JaegerSpan` into an interface based on what other code is consuming. This PR is not about fixing those parts, but rather, in reorganizing the existing classes.

pavolloffay replied:
> I extracted JaegerSpan into an interface based on what other code is consuming. This PR is not about fixing those parts, but rather, in reorganizing the existing classes.

There is a difference to what is needed in this repo(mostly tests) and what is exposed in general.

I don't see a value exposing a tracer in a span as the tracer is needed to start a span.

jpkrohling replied:
There's zero difference between what this PR exposes and what the current code exposes, from a consumer's perspective. I'm not exposing anything *more* than before: code that had `io.jaegertracing.Span` before already had access to `getTracer()`.

This PR is really just about reorganizing the structure of the existing classes and is already too big to my taste. I do plan on sending more PRs after this one, moving things we don't intend to expose, such as this `getTracer()` method.

Work on defining the public/internal API will be mostly done as part of #396.

pavolloffay commented on jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java:
```diff
@@ -0,0 +1,43 @@
+
+  long getDuration();
+
+  List<Reference> getReferences();
```
This is more suitable for span context

jpkrohling replied:
See my [previous comment](https://github.com/jaegertracing/jaeger-client-java/pull/454#discussion_r196121636)

pavolloffay commented on jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java:
```diff
@@ -0,0 +1,43 @@
+
+  String getOperationName();
+
+  String getServiceName();
```
This seems more suitable for a tracer

jpkrohling replied:
See my [previous comment](https://github.com/jaegertracing/jaeger-client-java/pull/454#discussion_r196121636)

pavolloffay commented on jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java:
```diff
@@ -0,0 +1,43 @@
+
+  long getStart();
+
+  long getDuration();
```
what are the units? How is defined duration for not finished span?

jpkrohling replied:
The interface does need some javadoc.

jpkrohling commented:
> I don't see much benefit of JaegerSpan interface. Could you please alaborate why interface and not just exposing public methods?

The interface is what we tell the world that they can rely on. 
JaegerBaseSpan is what we use internally. There might be public methods here which are *not* part of the public interface. For instance, the `getTracer()` method could be part of the `JaegerBaseSpan`, but not `JaegerSpan`.

mabn commented on jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java:
```diff
@@ -0,0 +1,68 @@
+  List<Reference> getReferences();
+
+  /**
+   * @return an immutable map of tags that are part of this span
```
It's not immutable at the moment - the content can change at any time.

jpkrohling replied:
Perhaps `unmodifiable` would be a better term, as implementations should return a copy that cannot be modified by the caller (like, `return Collections.unmodifiableMap(tags);`)

yurishkuro commented on jaeger-core/src/main/java/io/jaegertracing/reporters/Reporter.java:
```diff
@@ -14,7 +14,7 @@
 package io.jaegertracing.reporters;
 
-import io.jaegertracing.Span;
+import io.opentracing.Span;
```
I don't think this is a good direction. The only OpenTracing facade this library provides is the Tracer API. Reporter and everything else are internal components that should only work with internal span type, and the Tracer is the one responsible for downcasting from OT span to our span. In particular, for reporting we don't even need downcasting, because `JaegerSpan.finish()` calls `tracer.report(this)`, so reporter gets the internal span directly.

jpkrohling commented:
@pavolloffay said:
> `JaegerTracer` seems to be missing.

I was planning on sending another PR for that change, as this one is already too big to my taste, but as I'm blocked by this PR here, I decided to add another commit with this rename as well.

jpkrohling commented:
I just had a chat with @yurishkuro and we decided to revert the split between interface and implementation for JaegerSpan. It will be dealt with on #396, part of a bigger scope.

I updated this PR to include this change and also adjusted the PR description.

yurishkuro commented on jaeger-core/src/main/java/io/jaegertracing/baggage/BaggageSetter.java:
```diff
@@ -35,41 +35,42 @@ public BaggageSetter(BaggageRestrictionManager restrictionManager, Metrics metri
-      logFields(span, key, value, prevItem, truncated, restriction.isKeyAllowed());
-      return span.context();
+      logFields(jaegerSpan, key, value, prevItem, truncated, restriction.isKeyAllowed());
+      return ((JaegerSpanContext) jaegerSpan.context());
```
cast unnecessary

yurishkuro commented on jaeger-core/src/main/java/io/jaegertracing/baggage/BaggageSetter.java:
```diff
@@ -35,41 +35,42 @@ public BaggageSetter(BaggageRestrictionManager restrictionManager, Metrics metri
+    logFields(jaegerSpan, key, value, prevItem, truncated, restriction.isKeyAllowed());
     metrics.baggageUpdateSuccess.inc(1);
-    return span.context().withBaggageItem(key, value);
+    return ((JaegerSpanContext) jaegerSpan.context()).withBaggageItem(key, value);
```
cast unnecessary

yurishkuro commented on jaeger-core/src/main/java/io/jaegertracing/baggage/BaggageSetter.java:
```diff
@@ -35,41 +35,42 @@ public BaggageSetter(BaggageRestrictionManager restrictionManager, Metrics metri
-    if (!span.context().isSampled()) {
+  private void logFields(JaegerSpan jaegerSpan, String key, String value, String prevItem,
+                         boolean truncated, boolean valid) {
+    if (!((JaegerSpanContext) jaegerSpan.context()).isSampled()) {
```
cast unnecessary

yurishkuro commented on jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java:
```diff
@@ -71,13 +71,15 @@ private B3TextMapCodec(Builder builder) {
 
   @Override
   public void inject(SpanContext spanContext, TextMap carrier) {
```
as this is internal API, the spanContext should already be Jaeger's

yurishkuro commented on jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java:
```diff
@@ -14,9 +14,9 @@
-import io.jaegertracing.baggage.BaggageRestrictionManager;
+import io.jaegertracing.JaegerSpanContext;
+import io.jaegertracing.JaegerTracer;
+import io.opentracing.SpanContext;
```
this is a semantic change, previously methods were accepting Jaeger context

jpkrohling replied:
Reverted.

yurishkuro commented on jaeger-core/src/main/java/io/jaegertracing/propagation/CompositeCodec.java:
```diff
@@ -14,7 +14,7 @@
 package io.jaegertracing.propagation;
 
-import io.jaegertracing.SpanContext;
+import io.opentracing.SpanContext;
```
semantic change

yurishkuro commented on jaeger-core/src/main/java/io/jaegertracing/propagation/Extractor.java:
```diff
@@ -14,11 +14,11 @@
 package io.jaegertracing.propagation;
 
-import io.jaegertracing.SpanContext;
+import io.opentracing.SpanContext;
```
semantic change. Codecs are an internal API, Tracer takes care of converting the context to Jaeger context.

yurishkuro commented on jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java:
```diff
@@ -146,15 +148,15 @@ public void testTracerTagslist() {
     System.setProperty(Configuration.JAEGER_SERVICE_NAME, "Test");
     System.setProperty(Configuration.JAEGER_TAGS, "testTag1, testTag2 = testValue2");
-    io.jaegertracing.Tracer tracer = (io.jaegertracing.Tracer) Configuration.fromEnv().getTracer();
+    JaegerTracer tracer = (JaegerTracer) Configuration.fromEnv().getTracer();
```
TBD - I think configuration should be returning JaegerTracer. But it's not part of this PR.

yurishkuro commented on jaeger-core/src/test/java/io/jaegertracing/JaegerSpanTest.java:
```diff
@@ -79,56 +80,56 @@ public void testSpanMetrics() {
             .withBaggageRestrictionManager(mgr)
             .build();
-    span = (Span) tracer.buildSpan("some-operation").start();
+    jaegerSpan = (JaegerSpan) tracer.buildSpan("some-operation").start();
```
is this cast necessary?

jpkrohling replied:
Actually, we could mark JaegerTracer's `buildSpan()` to return `JaegerTracer.SpanBuilder`, whose `start()` method would return `JaegerSpan`. This way, callers using a concrete `JaegerTracer` won't need to cast anything...

jpkrohling replied:
Looks like there are other things to fix if we go this route, as `JaegerTracer.SpanBuilder` is not public, so, some tests cannot access it (`NoopMetricsFactoryTest` is the one that complained first).

I'll open an issue to investigate this.

yurishkuro commented on jaeger-core/src/main/java/io/jaegertracing/JaegerSpanContext.java:
```diff
@@ -16,12 +16,13 @@
 
 import io.jaegertracing.exceptions.EmptyTracerStateStringException;
 import io.jaegertracing.exceptions.MalformedTracerStateStringException;
+import io.opentracing.SpanContext;
```
Q: was the refactoring done using IDE type renaming, or via search/replace?  If the latter, it's hard to see if there are any unintended API changes in this diff since the previous type of `SpanContext` is now changed to OT via import, without showing up in the diff.

jpkrohling replied:
Initially, via the IDE, but I then changed some (most?) places to use OT's SpanContext. Based on our conversation, I later reverted this last step (also manually). The rule of thumb is that there should be no "semantic" change: where it was previously returning OT's Span Context should still return it and vice-versa.

yurishkuro commented on jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java:
```diff
@@ -137,7 +139,7 @@ public void testReporterConfigurationInvalidLogSpans() {
     System.setProperty(Configuration.JAEGER_SERVICE_NAME, "Test");
     System.setProperty(Configuration.JAEGER_TAGS, "testTag1=testValue1, testTag2 = testValue2");
-    io.jaegertracing.Tracer tracer = (io.jaegertracing.Tracer) Configuration.fromEnv().getTracer();
+    JaegerTracer tracer = (JaegerTracer) Configuration.fromEnv().getTracer();
```
cast still needed?

jpkrohling replied:
Yes, as `getTracer()` returns OT's Tracer and the test later calls `JaegerTracer#tags()`.

yurishkuro commented on jaeger-core/src/test/java/io/jaegertracing/baggage/BaggageSetterTest.java:
```diff
@@ -53,22 +54,22 @@ public void setUp() throws Exception {
             .withMetrics(metrics)
             .build();
-    span = (Span) tracer.buildSpan("some-operation").start();
+    jaegerSpan = (JaegerSpan) tracer.buildSpan("some-operation").start();
```
is the cast still necessary?

jpkrohling replied:
Yes, `tracer.buildSpan()` returns OT's SpanBuilder, which returns OT's Span for the `start()` method. It's basically the same as https://github.com/jaegertracing/jaeger-client-java/pull/454#discussion_r197764641

jpkrohling commented on jaeger-core/src/main/java/io/jaegertracing/Configuration.java:
```diff
@@ -217,7 +218,7 @@ public static Configuration fromEnv() {
   }
 
-  public synchronized io.opentracing.Tracer getTracer() {
+  public synchronized Tracer getTracer() {
```
Not on this PR: the previous one was OT's Tracer.

jpkrohling commented on jaeger-core/src/main/java/io/jaegertracing/JaegerTracer.java:
```diff
@@ -220,23 +223,24 @@ public void close() {
 
-      if (!(referencedContext instanceof SpanContext)) {
+      if (!(reference instanceof JaegerSpanContext)) {
         return this;
```
https://github.com/jaegertracing/jaeger-client-java/issues/466

objectiser commented on jaeger-core/src/main/java/io/jaegertracing/JaegerTracer.java:
```diff
@@ -220,23 +223,24 @@ public void close() {
 
     @Override
-    public io.opentracing.Tracer.SpanBuilder asChildOf(io.opentracing.SpanContext parent) {
+    public Tracer.SpanBuilder asChildOf(SpanContext parent) {
```
Should all the return types on the build be changed to `JaegerTracer.SpanBuilder`?
